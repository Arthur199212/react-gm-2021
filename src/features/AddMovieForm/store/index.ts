export * from './addMovieForm.selectors'
export * from './addMovieForm.slice'
export * from './addMovieForm.thunks'
