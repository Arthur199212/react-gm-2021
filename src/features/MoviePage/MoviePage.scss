@import 'variables';

.app-movie-page {
  .main {
    align-items: center;
    display: flex;
    flex-direction: column;

    .search-box-container {
      margin-bottom: 2rem;

      .search-header {
        width: 100%;

        .title {
          font-size: 2rem;
          font-weight: $bold;
          margin: 4rem auto 2rem;
          text-align: center;
        }
      }
    }
  }
}
