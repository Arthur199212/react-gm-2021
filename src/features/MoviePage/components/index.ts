export * from './Header'
export * from './MovieDetails'
