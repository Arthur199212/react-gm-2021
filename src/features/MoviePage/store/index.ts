export * from './moviePage.selectors'
export * from './moviePage.slice'
export * from './moviePage.thunks'
