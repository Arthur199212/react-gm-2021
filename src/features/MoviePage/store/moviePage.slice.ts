import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { Movie } from '@app/services/movies'

export enum MovieStatus {
  ERROR = 'error',
  LOADING = 'loading',
  SUCCESS = 'success'
}

export type MovieState = {
  isSsrData?: boolean
  movie?: Movie
  status: MovieStatus
}

export const initialState: MovieState = {
  status: MovieStatus.LOADING
}

const movieSlice = createSlice({
  name: 'movie',
  initialState,
  reducers: {
    setMovie: (state, { payload }: PayloadAction<Movie>) => {
      state.movie = payload
    },
    setSearchResult: (state, { payload }: PayloadAction<Movie>) => {
      state.movie = payload
      state.status = MovieStatus.SUCCESS
    },
    setStatus: (state, { payload }: PayloadAction<MovieStatus>) => {
      state.status = payload
    }
  }
})

export const movieReducer = movieSlice.reducer

export const { setMovie, setSearchResult, setStatus } = movieSlice.actions
