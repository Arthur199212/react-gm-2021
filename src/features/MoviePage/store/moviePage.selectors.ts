import { createSelector } from '@reduxjs/toolkit'
import { RootState } from '@app/store'
import { MovieState } from './moviePage.slice'

const selectMovieState = (state: RootState): MovieState => state.movie

export const selectMovie = createSelector(selectMovieState, state => state.movie)

export const selectMovieStatus = createSelector(selectMovieState, state => state.status)
