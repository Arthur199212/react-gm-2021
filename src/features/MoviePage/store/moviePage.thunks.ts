import { createAsyncThunk } from '@reduxjs/toolkit'
import { fetchMoviesByGenreThunk } from '@app/features/SearchPage/store'
import { moviesService } from '@app/services'
import { RootThunk } from '@app/store'
import { MovieStatus, setSearchResult, setStatus } from './moviePage.slice'

export const fetchMovieThunk = createAsyncThunk<void, string, RootThunk>(
  'movie/fetchMovieThunk',
  async (movieId, { dispatch }) => {
    dispatch(setStatus(MovieStatus.LOADING))

    try {
      const data = await moviesService.fetchMovie(movieId)

      // fetch movies with the same genre and wait for them
      await dispatch(fetchMoviesByGenreThunk(data.genres[0]))

      dispatch(setSearchResult(data))
    } catch (err) {
      dispatch(setStatus(MovieStatus.ERROR))
      console.error(err)
    }
  }
)
