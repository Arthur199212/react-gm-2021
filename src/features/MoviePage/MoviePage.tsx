import React from 'react'
import { Filters, Footer, SearchResults } from '@app/components'
import { Header, MovieDetails } from './components'

export const MoviePage = () => (
  <>
    <div className='app-movie-page content'>
      <Header />
      <main className='main'>
        <MovieDetails />
        <div className='container'>
          <Filters />
          <SearchResults />
        </div>
      </main>
    </div>
    <Footer />
  </>
)
