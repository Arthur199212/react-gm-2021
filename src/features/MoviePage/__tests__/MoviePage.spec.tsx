import React from 'react'
import { MoviePage } from '@app/features'
import { MovieStatus } from '@app/features/MoviePage/store'
import { INITIAL_APP_STATE, RootState } from '@app/store'
import { MOCK_MOVIE } from '@app/tests/mocks/mock-data'
import { render } from '@app/tests/testing-utils'

describe('MoviePage feature', () => {
  const state: RootState = {
    ...INITIAL_APP_STATE,
    movie: { movie: MOCK_MOVIE, status: MovieStatus.SUCCESS }
  }

  it('should render properly', () => {
    const { asFragment } = render(<MoviePage />, { state })

    expect(asFragment).toMatchSnapshot()
  })
})
