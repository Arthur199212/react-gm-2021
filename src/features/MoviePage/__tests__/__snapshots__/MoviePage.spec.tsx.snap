// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePage feature should render properly 1`] = `[Function]`;
